File: src/de/umr/fixcon/graphFunctions/standardFunctions/NegMaxDegreeFunction.java
package de.umr.fixcon.graphFunctions.standardFunctions;

import com.google.common.graph.Graph;
import de.umr.fixcon.graphFunctions.AbstractGraphFunction;
import de.umr.fixcon.graphFunctions.GraphFunction;

public class NegMaxDegreeFunction extends AbstractGraphFunction implements GraphFunction {
    @Override
    public boolean isEdgeMonotone() {
        return false;
    }

    @Override
    public double apply(Graph<Integer> g, final int... parameters) {
        return -1 * new MaxDegreeFunction().apply(g, parameters);
    }

    @Override
    public double optimum(int size) {
        return size > 2 ? -2 : -1;  //works if graph is connected
    }
}

File: src/de/umr/fixcon/graphFunctions/standardFunctions/NegMaxDegreeFunction.kt
package de.umr.fixcon.graphFunctions.standardFunctions

import de.umr.core.degreeSequence
import de.umr.fixcon.graphFunctions.GraphFunction
import org.jgrapht.Graph
import org.jgrapht.graph.DefaultEdge

/**Returns the maximum degree of all vertices in this graph.*/
class NegMaxDegreeFunction(k: Int) : GraphFunction(k = k) {

    override fun <V> completeAdditionBound(subgraph: Graph<V, DefaultEdge>) = 0

    override fun <V> eval(g: Graph<V, DefaultEdge>) = -g.degreeSequence.max()!!

    override fun globalOptimum() = -2
}